src/sagas/ui.js

```diff
@@ -37,6 +39,27 @@ function* githubExport(
   }
 }
 
+function* classroomExport(
+  successAction,
+  failureAction,
+  notDisplayedAction,
+  displayedAction) {
```

Comment:
If I’m not mistaken, the idea behind having the `githubExport` take all these parameters is so that it can agnostically handle both gist and repo exports. `classroomExport`, as written, is single-purpose, so there’s no need to parameterize the actions like this.

That said, I think it’s worth taking a step back and looking at the three “export” flows:

* Gist
* GitHub Repo
* Classroom

These three workflows are basically all identical, with the only variance being in what exactly goes into generating the export URL (which is of course a pretty complicated question involving various internal logic and calls to external APIs, but still, just one question).

Right now the flow is simply duplicated three times in the code, with duplicate actions, reducer logic, etc. I am generally a fan of a “rule of three” when it comes to code duplication: once you have three distinct code paths basically all doing the same thing, it’s time to refactor. This would be the point where the rule of three is triggered (I think snapshots are sufficiently different from the others to not be eligible).

So, I would propose restructuring the redux state/flow around a generic concept of an “export”, with action payloads including an export type, either `gist`, `repo`, or `classroom`. If I’m not mistaken, the only parts of the code that will need to be sensitive to the export type are:

* The `clients` saga
* The notification component (and probably only in that the translation key used is derived from the export type)

What do you think? Obviously that does add significant extra work to getting this feature released, but ultimately I think this is the right time to do the refactor (happy to justify that belief over Slack) and I think you’ll find that the total footprint of this PR won’t change substantially (there will just be more changed LOC and fewer new ones). Let me know!

Reply:
Hey @outoftime Yes this makes sense. I think it sounds like the right time to do it and ultimately more organized in the long run. Looking forward to implementing this restructuring and getting feedback.